File: day_02/p1.py
from typing import List


def parse_input(fname: str) -> List[List[int]]:
    lines = []
    with open(fname, 'r') as f:
        lines = f.read().splitlines()
    
    return [list(map(int, line.split())) for line in lines]

def is_safe_by_order_and_difference(record: List[int]) -> bool:
    
    diffs = [a-b for (a, b) in zip(record, record[1:])]

    return all(1 <= diff <= 3 for diff in diffs) or \
           all(-3 <= diff <= -1 for diff in diffs)

def compute(records: List[List[int]]) -> int:
    return sum(is_safe_by_order_and_difference(record) for record in records)

def main():
    records = parse_input("input.txt")
    print(compute(records))

if __name__ == "__main__":
    main()

File: day_02/p1.cpp
#include "pypp.hpp" // https://github.com/A-Bergonzoli/pypp/tree/main

#include <iostream>
#include <numeric>

using ints = std::vector<int>;

std::vector<ints> ParseInput(const std::string& fname)
{
    std::vector<ints> records {};
    const auto lines = pypp::splitFileLines(fname);

    for (const auto& line : lines) {
        ints record {};
        auto record_str = pypp::split(line, ' ');

        std::ranges::transform(record_str,
            std::back_inserter(record),
            [](const std::string& s) { return std::stoi(s); });

        records.push_back(record);
    }

    return records;
}

bool IsSafeByOrderAndDifference(const ints& record)
{
    ints diffs {};

    for (const auto& tuple : pypp::zip(record.begin() + 1, record.end(), record.begin()))
        diffs.push_back(std::get<1>(tuple) - std::get<0>(tuple));

    return std::ranges::all_of(diffs, [](const int diff) { return 1 <= diff && diff <= 3; })
        || std::ranges::all_of(diffs, [](const int diff) { return -3 <= diff && diff <= -1; });
}

int Compute(const std::vector<ints>& records)
{
    return std::accumulate(
        records.begin(),
        records.end(),
        0,
        [](int sum, const ints& record) { return sum + IsSafeByOrderAndDifference(record); });
}

int main()
{
    auto records = ParseInput("input.txt");
    std::cout << Compute(records) << std::endl;

    return 0;
}
